pkg/danm/danm.go

```diff
@@ -183,72 +189,98 @@ func extractConnections(args *cniArgs) error {
+  checkpoint, err := checkpoint_utils.GetCheckpoint()
+  if err != nil {
+    return nil, errors.New("failed to instantiate checkpoint object due to:" + err.Error())
   }
-  err := syncher.GetAggregatedResult()
-  return syncher.MergeCniResults(), err
+  resourceMap, err := checkpoint.GetComputeDeviceMap(string(args.podUid))
+  if err != nil || len(resourceMap) == 0 {
+    return nil, errors.New("failed to retrieve Pod info from checkpoint object due to:" + err.Error())
+  }
+  allocatedDevices := make(map[string][]string)
+  for resourcename, resources := range resourceMap {
+    if strings.HasPrefix(resourcename, devicePool) {
+      hostDevice, err := sriov_utils.GetPfName(resources.DeviceIDs[0])
+      if err != nil {
+        return nil, errors.New("failed to get the name of the sriov PF for device "+ resources.DeviceIDs[0] +" due to:" + err.Error())
+      }
+      allocatedDevices[hostDevice] = resources.DeviceIDs
+    }
+  }
+  return allocatedDevices, nil
 }
 
-func createInterface(syncher *syncher.Syncher, iface danmtypes.Interface, args *cniArgs) {
-  danmClient, err := createDanmClient(args.stdIn)
-  if err != nil {
-    syncher.PushResult(iface.Network, err, nil)
-    return
+func getDevice(hostDevice string, allocatedDevices map[string][]string)(string, map[string][]string) {
+  device :=  ""
+  if len(allocatedDevices[hostDevice]) != 0 {
+    device, allocatedDevices[hostDevice] = allocatedDevices[hostDevice][len(allocatedDevices[hostDevice])-1], allocatedDevices[hostDevice][:len(allocatedDevices[hostDevice])-1]
   }
-  isDelegationRequired, netInfo, err := cnidel.IsDelegationRequired(danmClient, iface.Network, args.nameSpace)
+  return device, allocatedDevices
+}
+
+func setupNetworking(args *cniArgs) (*current.Result, error) {
+  danmClient, err := createDanmClient(args.stdIn)
   if err != nil {
-    syncher.PushResult(iface.Network, err, nil)
-    return
+    return nil, errors.New("failed to create DanmClient due to:" + err.Error())
   }
+  syncher := syncher.NewSyncher(len(args.interfaces))
+  allocatedDevices := make(map[string][]string)
   var cniRes *current.Result
-  if isDelegationRequired {
-    cniRes, err = createDelegatedInterface(danmClient, iface, netInfo, args)
+  for nicID, nicParams := range args.interfaces {
+    isDelegationRequired, netInfo, err := cnidel.IsDelegationRequired(danmClient, nicParams.Network, args.nameSpace)
     if err != nil {
-      syncher.PushResult(iface.Network, err, nil)
-      return
+      return cniRes, errors.New("failed to get DanmNet due to:" + err.Error())
     }
-  } else {
-    cniRes, err = createDanmInterface(danmClient, iface, netInfo, args)
-    if err != nil {
-      syncher.PushResult(iface.Network, err, nil)
-      return
+    nicParams.DefaultIfaceName = "eth" + strconv.Itoa(nicID)
+    if isDelegationRequired {
+      if cnidel.IsDeviceNeeded(netInfo.Spec.NetworkType) {
+        if len(allocatedDevices) == 0 {
```

Comment:
so, if we have filled up the allocatedDevices map because of a DanmNet which had ties to DeviceA, we won't read out DeviceB of DanmNet2?

correct logic would be filling up allocatedDevices[netinfo.Spec.Options.DevicePool] if this particular record is empty

Reply:
Applied